src/loki_cluster.py

```diff
@@ -0,0 +1,321 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+
+"""This module contains an endpoint wrapper class for the requirer side of the ``loki-cluster`` relation.
+
+As this relation is cluster-internal and not intended for third-party charms to interact with `loki-coordinator-k8s`, its only user will be the loki-worker-k8s charm. As such, it does not live in a charm lib as most other relation endpoint wrappers do.
+"""
+
+import json
+import logging
+from enum import Enum, unique
+from typing import Any, Dict, Iterable, List, MutableMapping, Optional, Set
+from urllib.parse import urlparse
+
+import ops
+import pydantic
+from ops import EventSource, Object, ObjectEvents, RelationCreatedEvent
+from pydantic import BaseModel, ConfigDict
+
+log = logging.getLogger("loki_cluster")
+
+DEFAULT_ENDPOINT_NAME = "loki-cluster"
+BUILTIN_JUJU_KEYS = {"ingress-address", "private-address", "egress-subnets"}
+LOKI_CONFIG_FILE = "/etc/loki/loki-config.yaml"
+LOKI_CERT_FILE = "/etc/loki/server.cert"
+LOKI_KEY_FILE = "/etc/loki/private.key"
+LOKI_CLIENT_CA_FILE = "/etc/loki/ca.cert"
+
+
+# TODO: inherit enum.StrEnum when jammy is no longer supported.
+# https://docs.python.org/3/library/enum.html#enum.StrEnum
+@unique
+class LokiRole(str, Enum):
+    """Loki component role names."""
+
+    # Meta roles based on:
+    # https://grafana.com/docs/loki/latest/get-started/deployment-modes/#simple-scalable
+    read = "read"
+    write = "write"
+    backend = "backend"
+    all = "all"
+
+
+META_ROLES = {}
+
+
+def expand_roles(roles: Iterable[LokiRole]) -> Set[LokiRole]:
+    """Expand any meta roles to their 'atomic' equivalents."""
+    expanded_roles = set()
+    for role in roles:
+        if role in META_ROLES:
+            expanded_roles.update(META_ROLES[role])
+        else:
+            expanded_roles.add(role)
+    return expanded_roles
+
+
+class ConfigReceivedEvent(ops.EventBase):
+    """Event emitted when the "loki-cluster" provider has shared a new loki config."""
+
+    config: Dict[str, Any]
```

Comment:
not sure we need this as class attribute, why is it here?

Reply:
or is it just docs?

Reply:
It's a leftover from mimir-worker. It's not necessary.